hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala

```diff
@@ -19,86 +19,121 @@ object TableStage {
     globals: IR,
     partitioner: RVDPartitioner,
     contexts: IR,
-    body: (IR) => IR): TableStage = {
-    val globalsID = genUID()
-    new TableStage(
-      FastIndexedSeq(globalsID -> globals),
-      Set(globalsID),
-      Ref(globalsID, globals.typ),
+    body: (Ref) => IR
+  ): TableStage = {
+    val globalsRef = Ref(genUID(), globals.typ)
+    TableStage(
+      FastIndexedSeq(globalsRef.name -> globals),
+      FastIndexedSeq(globalsRef.name -> globalsRef),
+      globalsRef,
       partitioner,
-      contexts) {
-      def partition(ctxRef: Ref): IR = body(ctxRef)
-    }
+      contexts,
+      body)
+  }
+
+  def apply(
+    letBindings: IndexedSeq[(String, IR)],
+    broadcastVals: IndexedSeq[(String, IR)],
+    globals: Ref,
+    partitioner: RVDPartitioner,
+    contexts: IR,
+    partition: Ref => IR
+  ): TableStage = {
+    val ctxType = contexts.typ.asInstanceOf[TStream].elementType
+    val ctxRef = Ref(genUID(), ctxType)
+
+    new TableStage(letBindings, broadcastVals, globals, partitioner, contexts, ctxRef.name, partition(ctxRef))
   }
 }
 
-abstract class TableStage(
+// Scope structure:
+// * 'letBindings' are evaluated in scope of previous 'letBindings', and are
+//   visible in 'broadcastVals' and 'contexts'.
+// * 'broadcastVals' are evaluated in scope of 'letBindings', and are visible
+//   in 'partitionIR'.
+// * 'globals' must be bound in 'letBindings', and rebound in 'broadcastVals',
+//   so 'globals' is visible both in later 'letBindings' and in 'partitionIR'.
+class TableStage(
   val letBindings: IndexedSeq[(String, IR)],
-  val broadcastVals: Set[String],
-  val globals: IR,
+  val broadcastVals: IndexedSeq[(String, IR)],
+  val globals: Ref,
   val partitioner: RVDPartitioner,
-  val contexts: IR) { self =>
-
-  def partition(ctxRef: Ref): IR
-
-  def wrapInBindings(body: IR): IR = {
-    letBindings.foldRight(body) { case ((name, binding), soFar) => Let(name, binding, soFar) }
+  val contexts: IR,
+  val ctxRefName: String,
+  val partitionIR: IR) { self =>
```

Comment:
this should probably be private.

Reply:
`def partition(ctx)` should be the interface consumers use to construct the partition stream.